File: src/zotero.rs
use rusqlite::{Connection, OpenFlags};
use serde::{Deserialize, Serialize};
use std::env;
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ZoteroItem {
    pub item_id: i64,
    pub key: String,
    pub title: Option<String>,
    pub authors: Option<String>,
    pub date: Option<String>,
    pub abstract_note: Option<String>,
    pub pdf_path: Option<String>,
}

pub struct ZoteroReader {
    db_path: PathBuf,
    storage_dir: PathBuf,
}

impl ZoteroReader {
    /// Try to find the Zotero database path.
    /// Priority:
    /// 1. ZOTERO_DB_PATH environment variable (if set)
    /// 2. Default location: ~/Zotero/zotero.sqlite
    fn find_db_path() -> anyhow::Result<PathBuf> {
        // First, check if ZOTERO_DB_PATH is explicitly set
        if let Ok(db_path_str) = env::var("ZOTERO_DB_PATH") {
            let db_path = PathBuf::from(&db_path_str);
            if db_path.exists() {
                return Ok(db_path);
            }
        }

        if let Some(home_dir) = dirs::home_dir() {
            let default_path = home_dir.join("Zotero").join("zotero.sqlite");
            if default_path.exists() {
                return Ok(default_path);
            }
        }

        Err(anyhow::anyhow!(
            "Could not find Zotero database. Either set ZOTERO_DB_PATH environment variable or ensure Zotero is installed at the default location (~/Zotero/zotero.sqlite)"
        ))
    }

    pub fn new() -> anyhow::Result<Self> {
        let db_path = Self::find_db_path()?;

        // Storage dir is in the same parent directory as the DB, under "storage"
        let storage_dir = db_path.parent().unwrap().join("storage");

        Ok(Self {
            db_path,
            storage_dir,
        })
    }

    pub fn get_items(&self) -> anyhow::Result<Vec<ZoteroItem>> {
        // Open in read-only mode
        let conn = Connection::open_with_flags(
            &self.db_path,
            OpenFlags::SQLITE_OPEN_READ_ONLY | OpenFlags::SQLITE_OPEN_URI,
        )?;

        // Query to get items. We'll join to get basic metadata.
        // Zotero field IDs: 1=title, 2=abstractNote, 14=date (publication date)
        let mut stmt = conn.prepare(
            "SELECT 
                i.itemID,
                i.key,
                title_val.value as title,
                abstract_val.value as abstract,
                date_val.value as date,
                (
                    SELECT GROUP_CONCAT(
                        CASE 
                            WHEN c.firstName IS NOT NULL AND c.lastName IS NOT NULL 
                            THEN c.lastName || ', ' || c.firstName
                            WHEN c.lastName IS NOT NULL 
                            THEN c.lastName
                            ELSE NULL
                        END, '; '
                    )
                    FROM itemCreators ic
                    JOIN creators c ON ic.creatorID = c.creatorID
                    WHERE ic.itemID = i.itemID
                ) as authors
            FROM items i
            JOIN itemTypes it ON i.itemTypeID = it.itemTypeID
            -- Get title (fieldID = 1)
            LEFT JOIN itemData title_data ON i.itemID = title_data.itemID AND title_data.fieldID = 1
            LEFT JOIN itemDataValues title_val ON title_data.valueID = title_val.valueID
            -- Get abstract (fieldID = 2)
            LEFT JOIN itemData abstract_data ON i.itemID = abstract_data.itemID AND abstract_data.fieldID = 2
            LEFT JOIN itemDataValues abstract_val ON abstract_data.valueID = abstract_val.valueID
            -- Get publication date (fieldID = 14)
            LEFT JOIN itemData date_data ON i.itemID = date_data.itemID AND date_data.fieldID = 14
            LEFT JOIN itemDataValues date_val ON date_data.valueID = date_val.valueID
            
            WHERE it.typeName NOT IN ('attachment', 'note', 'annotation')
            ORDER BY i.dateModified DESC"
        )?;

        let item_iter = stmt.query_map([], |row| {
            Ok(ZoteroItem {
                item_id: row.get(0)?,
                key: row.get(1)?,
                title: row.get(2).ok(),
                abstract_note: row.get(3).ok(),
                date: row.get(4).ok(),
                authors: row.get(5).ok(),
                pdf_path: None, // Will be filled later
            })
        })?;

        let mut items = Vec::new();
        for item in item_iter {
            items.push(item?);
        }

        // Now find attachments for each item
        // We can do this in a separate query or loop.
        // For efficiency, maybe we query all PDF attachments and map them.
        let mut attachment_stmt = conn.prepare(
            "SELECT ia.parentItemID, ia.path, att.key 
             FROM itemAttachments ia
             JOIN items att ON att.itemID = ia.itemID
             WHERE ia.contentType = 'application/pdf'",
        )?;

        let attachment_iter = attachment_stmt.query_map([], |row| {
            Ok((
                row.get::<_, Option<i64>>(0)?,
                row.get::<_, Option<String>>(1)?,
                row.get::<_, String>(2)?,
            ))
        })?;

        let mut pdf_map = std::collections::HashMap::new();
        for res in attachment_iter {
            let (parent_id_opt, path_opt, key) = res?;
            let Some(parent_id) = parent_id_opt else {
                // Zotero can contain standalone/linked attachments without a parent item.
                // Those cannot be associated to a bibliographic item, so we skip them.
                continue;
            };
            if let Some(path_str) = path_opt {
                pdf_map.insert(parent_id, (path_str, key));
            }
        }

        for item in &mut items {
            if let Some((path_str, key)) = pdf_map.get(&item.item_id) {
                // Resolve path
                if path_str.starts_with("storage:") {
                    let rel_path = path_str.trim_start_matches("storage:");
                    let full_path = self.storage_dir.join(key).join(rel_path);
                    if full_path.exists() {
                        item.pdf_path = Some(full_path.to_string_lossy().to_string());
                    }
                } else {
                    // Absolute path?
                    let p = PathBuf::from(path_str);
                    if p.exists() {
                        item.pdf_path = Some(path_str.clone());
                    }
                }
            }
        }

        Ok(items)
    }
}

File: src/litlake/zotero.py
from __future__ import annotations

import os
import sqlite3
from dataclasses import dataclass, field
from pathlib import Path

from litlake.providers.extraction import SUPPORTED_EXTRACTION_MIME_TYPES


@dataclass
class ZoteroAttachment:
    item_id: int
    key: str
    parent_item_id: int
    content_type: str
    path: str
    resolved_path: str
    link_mode: int | None = None


@dataclass
class ZoteroAnnotation:
    item_id: int
    key: str
    parent_item_id: int
    parent_attachment_item_id: int
    parent_attachment_key: str | None
    annotation_type: int
    author_name: str | None
    text: str | None
    comment: str | None
    color: str | None
    page_label: str | None
    sort_index: str
    position: str
    is_external: int


@dataclass
class ZoteroNote:
    item_id: int
    key: str
    parent_item_id: int
    note_html: str | None
    title: str | None


@dataclass
class ZoteroItem:
    item_id: int
    key: str
    title: str | None
    authors: str | None
    date: str | None
    abstract_note: str | None
    attachments: list[ZoteroAttachment] = field(default_factory=list)
    annotations: list[ZoteroAnnotation] = field(default_factory=list)
    notes: list[ZoteroNote] = field(default_factory=list)


class ZoteroReader:
    def __init__(self, db_path: str | None = None):
        self.db_path = self._find_db_path(db_path)
        self.storage_dir = self.db_path.parent / "storage"

    @staticmethod
    def _find_db_path(explicit: str | None = None) -> Path:
        candidate = explicit or os.getenv("ZOTERO_DB_PATH")
        if candidate and "${" not in candidate:
            path = Path(candidate).expanduser().resolve()
            if path.exists():
                return path

        default_path = (Path.home() / "Zotero" / "zotero.sqlite").resolve()
        if default_path.exists():
            return default_path

        raise FileNotFoundError(
            "Could not find Zotero database. Set ZOTERO_DB_PATH or use default ~/Zotero/zotero.sqlite"
        )

    def _resolve_attachment_path(self, path_str: str | None, attachment_key: str) -> str | None:
        if not path_str:
            return None
        if path_str.startswith("storage:"):
            rel_path = path_str.replace("storage:", "", 1)
            full_path = (self.storage_dir / attachment_key / rel_path).resolve()
        else:
            full_path = Path(path_str).expanduser().resolve()
        if not full_path.exists():
            return None
        return str(full_path)

    def get_items(self) -> list[ZoteroItem]:
        conn = sqlite3.connect(f"file:{self.db_path}?mode=ro", uri=True)
        conn.row_factory = sqlite3.Row

        query = """
            SELECT
                i.itemID,
                i.key,
                title_val.value as title,
                abstract_val.value as abstract,
                SUBSTR(date_val.value, 1, 4) as date,
                (
                    SELECT GROUP_CONCAT(
                        CASE
                            WHEN c.firstName IS NOT NULL AND c.lastName IS NOT NULL
                            THEN c.lastName || ', ' || c.firstName
                            WHEN c.lastName IS NOT NULL
                            THEN c.lastName
                            ELSE NULL
                        END, '; '
                    )
                    FROM itemCreators ic
                    JOIN creators c ON ic.creatorID = c.creatorID
                    WHERE ic.itemID = i.itemID
                ) as authors
            FROM items i
            JOIN itemTypes it ON i.itemTypeID = it.itemTypeID
            LEFT JOIN itemData title_data ON i.itemID = title_data.itemID AND title_data.fieldID = 1
            LEFT JOIN itemDataValues title_val ON title_data.valueID = title_val.valueID
            LEFT JOIN itemData abstract_data ON i.itemID = abstract_data.itemID AND abstract_data.fieldID = 2
            LEFT JOIN itemDataValues abstract_val ON abstract_data.valueID = abstract_val.valueID
            LEFT JOIN itemData date_data ON i.itemID = date_data.itemID AND date_data.fieldID = 6
            LEFT JOIN itemDataValues date_val ON date_data.valueID = date_val.valueID
            WHERE it.typeName NOT IN ('attachment', 'note', 'annotation')
            ORDER BY i.dateModified DESC
        """

        rows = conn.execute(query).fetchall()
        items = [
            ZoteroItem(
                item_id=int(r[0]),
                key=r[1],
                title=r[2],
                abstract_note=r[3],
                date=r[4],
                authors=r[5],
            )
            for r in rows
        ]
        items_by_id = {item.item_id: item for item in items}

        attachment_placeholders = ",".join("?" for _ in SUPPORTED_EXTRACTION_MIME_TYPES)
        attachment_rows = conn.execute(
            f"""
            SELECT
                ia.itemID,
                ia.parentItemID,
                ia.linkMode,
                ia.contentType,
                ia.path,
                att.key
            FROM itemAttachments ia
            JOIN items att ON att.itemID = ia.itemID
            WHERE ia.contentType IN ({attachment_placeholders})
            """,
            tuple(sorted(SUPPORTED_EXTRACTION_MIME_TYPES)),
        ).fetchall()

        for attachment_row in attachment_rows:
            attachment_item_id = int(attachment_row[0])
            parent_item_id = attachment_row[1]
            link_mode = attachment_row[2]
            content_type = attachment_row[3]
            path_str = attachment_row[4]
            attachment_key = attachment_row[5]
            if parent_item_id is None or path_str is None or content_type is None:
                continue
            item = items_by_id.get(int(parent_item_id))
            if item is None:
                continue
            resolved_path = self._resolve_attachment_path(path_str, attachment_key)
            if resolved_path is None:
                continue
            item.attachments.append(
                ZoteroAttachment(
                    item_id=attachment_item_id,
                    key=attachment_key,
                    parent_item_id=int(parent_item_id),
                    content_type=str(content_type),
                    path=str(path_str),
                    resolved_path=resolved_path,
                    link_mode=int(link_mode) if link_mode is not None else None,
                )
            )

        annotation_rows = conn.execute(
            """
            SELECT
                ann_item.itemID,
                ann_item.key,
                att.parentItemID,
                ia.parentItemID,
                parent_att.key,
                ia.type,
                ia.authorName,
                ia.text,
                ia.comment,
                ia.color,
                ia.pageLabel,
                ia.sortIndex,
                ia.position,
                ia.isExternal
            FROM itemAnnotations ia
            JOIN items ann_item ON ann_item.itemID = ia.itemID
            LEFT JOIN itemAttachments att ON att.itemID = ia.parentItemID
            LEFT JOIN items parent_att ON parent_att.itemID = ia.parentItemID
            """
        ).fetchall()

        for row in annotation_rows:
            parent_item_id = row[2]
            if parent_item_id is None:
                continue
            item = items_by_id.get(int(parent_item_id))
            if item is None:
                continue
            item.annotations.append(
                ZoteroAnnotation(
                    item_id=int(row[0]),
                    key=row[1],
                    parent_item_id=int(parent_item_id),
                    parent_attachment_item_id=int(row[3]) if row[3] is not None else 0,
                    parent_attachment_key=row[4],
                    annotation_type=int(row[5]) if row[5] is not None else 0,
                    author_name=row[6],
                    text=row[7],
                    comment=row[8],
                    color=row[9],
                    page_label=row[10],
                    sort_index=row[11] or "",
                    position=row[12] or "",
                    is_external=int(row[13]) if row[13] is not None else 0,
                )
            )

        note_rows = conn.execute(
            """
            SELECT
                note_item.itemID,
                note_item.key,
                inote.parentItemID,
                inote.note,
                inote.title
            FROM itemNotes inote
            JOIN items note_item ON note_item.itemID = inote.itemID
            """
        ).fetchall()

        for row in note_rows:
            parent_item_id = row[2]
            if parent_item_id is None:
                continue
            item = items_by_id.get(int(parent_item_id))
            if item is None:
                continue
            item.notes.append(
                ZoteroNote(
                    item_id=int(row[0]),
                    key=row[1],
                    parent_item_id=int(parent_item_id),
                    note_html=row[3],
                    title=row[4],
                )
            )

        conn.close()
        return items


__all__ = [
    "ZoteroAttachment",
    "ZoteroAnnotation",
    "ZoteroItem",
    "ZoteroNote",
    "ZoteroReader",
]
